File: Data Structures & Algorithms/reorder-linked-list/submission-0.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

class Solution {
    public void reorderList(ListNode head) {
        if (head == null || head.next == null) {
            return;
        }
        List<ListNode> nodes = new ArrayList<>();
        ListNode cur = head;
        while (cur != null) {
            nodes.add(cur);
            cur = cur.next;
        }
        int p1 = 0;
        int p2 = nodes.size() - 1;
        while (p1 < p2){
            nodes.get(p1).next = nodes.get(p2);
            nodes.get(p2).next = nodes.get(p1+1);
            p1++;
            p2--;
        }
        nodes.get(p1).next = null;
    }
}

File: Data Structures & Algorithms/remove-node-from-end-of-linked-list/submission-0.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        pre = ListNode(0)
        pre.next = head
       
        temp = head
     
        temp_pre = pre

        number = 0
        while temp:
            temp = temp.next
            number += 1
        m = number - n + 1
  
        while m > 1:
            temp_pre = temp_pre.next
            m -= 1

        temp_pre.next = temp_pre.next.next
     
        return pre.next



        